### ISSUE TITLE
On arm64, "hugo -s docs" fails with "function not implemented"

### ISSUE THREAD

**Issue by anthonyfok** [spf13/hugo#1772]
According to the Debian arm64 build daemon [build log](https://buildd.debian.org/status/fetch.php?pkg=hugo&arch=arm64&ver=0.15%2Bgit20160109.147.dd1d655-1&stamp=1452558812), a `hugo -s docs` would fail with "function not implemented" error:

```
# TODO: Add "hugo-docs" package in the future
obj-aarch64-linux-gnu/bin/hugo -s docs -d public/html --uglyURLs
2016/01/12 00:33:24 function not implemented
debian/rules:9: recipe for target 'override_dh_auto_build' failed
```

Initial code search on "function not implemented" led me to `golang.org/x/sys/unix`, Errno 38 or 0x26, i.e., `ENOSYS`.

I was able to reproduce this on Debian arm64 developer machine asachi.debian.org, and it is not one single "function not implemented" errors, but random amount of it, from 9 errors to 94812 errors.  I'll need more time to investigate.

Note that this problem does **not** affect `armhf` and `armel` platforms.

**Comment by anthonyfok** [comment]
After adding

```go
log.SetFlags(log.LstdFlags | log.Lshortfile)
```

to Hugo’s main.go (credit: http://stackoverflow.com/questions/24809287/how-do-you-get-a-go-golang-program-to-print-the-line-number-of-the-error-it-ju), the source of the error became more apparent:

```
2016/01/13 04:17:32 viper.go:260: error: function not implemented
2016/01/13 04:17:32 viper.go:260: error: function not implemented
2016/01/13 04:17:32 viper.go:260: error: function not implemented
2016/01/13 04:17:33 viper.go:260: error: function not implemented
2016/01/13 04:17:33 viper.go:260: error: function not implemented
2016/01/13 04:17:33 viper.go:260: error: function not implemented
2016/01/13 04:17:33 viper.go:260: error: function not implemented
2016/01/13 04:17:33 viper.go:260: error: function not implemented
0 draft content
0 future content
165 pages created
0 non-page files copied
0 paginator pages created
41 tags created
0 groups created
in 1100 ms
```

The error messages were printed by `log.Println("error:", err)` on [Line 260](https://github.com/spf13/viper/blob/a212099cbe6fbe8d07476bfda8d2d39b6ff8f325/viper.go#L260) inside a [goroutine](https://tour.golang.org/concurrency/1), hence the multiple and random numbers of errors.  However, it does continue and finish rendering the site.

And, in the case of the current Debian Hugo package, which was compiled with an older viper (commit dated 20151224) and older fsnotify (1.2.1 vs the current 1.2.8), the error was printed a bit earlier, by `log.Fatal(error)` on [line 236](https://github.com/spf13/viper/blob/110492b3001269a0777997fcbe68a60828fac4cb/viper.go#L234-L237), hence the one-shot error message and immediate exit:

```
2016/01/13 04:21:53 viper.go:236: function not implemented
```

Nevertheless, they both happen inside viper.WatchConfig() (hence [fsnotify](https://github.com/go-fsnotify/fsnotify)), which is called from either commands/hugo.go or commands/server.go, depending on which Hugo command is run.

I will look inside [fsnotify](https://github.com/go-fsnotify/fsnotify) next, as I suspect fsnotify could be using a syscall function that is not available on Linux/arm64 for one reason or another, like in https://github.com/golang/go/issues/11918.

Anyhow, this does bring up a question:
@spf13, should commands/hugo.go be running watchConfig() even when it is _not_ running in watch mode?

**Comment by anthonyfok** [comment]
It turns out the "function not implemented" (ENOSYS) errors originated from [Line 85 in inotify_poller.go](https://github.com/go-fsnotify/fsnotify/blob/master/inotify_poller.go#L85) from fsnotify:

```go
	n, errno := syscall.EpollWait(poller.epfd, events, -1)
```

Why?  Because epoll_wait is [deprecated](http://people.linaro.org/~rikuvoipio/aarch64-talk/?transition=linear#/18/1) according to a slideshow by Linaro who ported Linux to ARM64, and therefore, only the newer epoll_pwait is provided on Linux/arm64.

Problem is, Go has not yet implemented the corresponding EpollPwait, see the "Unimplemented" section of [go/src/syscall/syscall_linux.go:948](https://go.googlesource.com/go/+/master/src/syscall/syscall_linux.go#948) and [golang.org/x/sys/unix/syscall_linux.go:971](https://go.googlesource.com/sys/+/master/unix/syscall_linux.go#971).

It seems that the Go developers are aware of the issue, see golang/go#10235, though the milestone there changed from **Go1.5** to **Go1.6Early**, and now **Unplanned**...  Oh well...  We'll see what happens next.

In the meantime, if anyone were really using Hugo on ARM64, we can happily report that all features work except `--watch`.  :wink:

**Comment by anthonyfok** [comment]
Looks like it has been fixed in golang.org/x/sys/unix, thanks to Riku Voipio (@suihkulokki) in commit golang/sys@324e137, and the upstream golang/go#10235 has closed accordingly.

I would like to refresh relevant Go packages in Debian and make sure it is all working properly on arm64 test machine before coming back and close this issue. :-)

**Comment by anthonyfok** [comment]
Fixed by Riku Voipio in golang/sys@324e137 and fsnotify/fsnotify#135 (committed by Nathan Youngman).
When compiled with the latest fsnotify v1.3.0, `hugo serve` (with `--watch`) works great on Linux/arm64 (aarch64).  :+1:

Thank you all!

**Comment by spf13** [comment]
@anthonyfok we should add this to the release notes for 0.16
